src/uct/cuda/cuda_ipc/cuda_ipc_ep.c

```diff
@@ -31,16 +35,151 @@ UCS_CLASS_DEFINE_NEW_FUNC(uct_cuda_ipc_ep_t, uct_ep_t, uct_iface_t*,
                           const uct_device_addr_t *, const uct_iface_addr_t *);
 UCS_CLASS_DEFINE_DELETE_FUNC(uct_cuda_ipc_ep_t, uct_ep_t);
 
-ucs_status_t uct_cuda_ipc_ep_get_zcopy(uct_ep_h tl_ep, const uct_iov_t *iov, size_t iovcnt,
-                                       uint64_t remote_addr, uct_rkey_t rkey,
-                                       uct_completion_t *comp)
+#define uct_cuda_ipc_trace_data(_remote_addr, _rkey, _fmt, ...)     \
+    ucs_trace_data(_fmt " to %"PRIx64"(%+ld)", ## __VA_ARGS__, (_remote_addr), \
+                   (_rkey))
+
+#define UCT_CUDA_IPC_ZERO_LENGTH_POST(len)                      \
+    do {                                                        \
+        if (0 == len) {                                         \
+            ucs_trace_data("Zero length request: skip it");     \
+            return UCS_OK;                                      \
+        }                                                       \
+    } while(0);
+
+static UCS_F_ALWAYS_INLINE ucs_status_t
+uct_cuda_ipc_get_mapped_addr(uct_cuda_ipc_key_t *key, int cu_device,
+                             uint64_t remote_addr, void *mapped_rem_addr,
+                             const uct_iov_t *iov, uct_cuda_ipc_ep_t *ep,
+                             uct_cuda_ipc_iface_t *iface)
 {
+    void *mapped_rem_base_addr = NULL;
+    int offset = 0;
+    CUresult cu_err;
+    CUresult cu_ret;
+    CUcontext local_ptr_ctx;
+    CUcontext remote_ptr_ctx;
+    CUpointer_attribute  attribute;
+
+    if (key->dev_num == (int) cu_device) {
+        attribute = CU_POINTER_ATTRIBUTE_CONTEXT;
+        cu_ret    = cuPointerGetAttribute((void *) &remote_ptr_ctx, attribute,
+                                          (CUdeviceptr) remote_addr);
+        if (CUDA_ERROR_INVALID_VALUE != cu_ret) {
+            /* another process's contex using the same device */
+            cu_err = cu_ret;
+        } else if (cu_ret != CUDA_SUCCESS) {
+            cuGetErrorString(cu_ret, &cu_err_str);
+            ucs_error("cuPointerGetAttribute failed ret:%s", cu_err_str);
+            return UCS_ERR_IO_ERROR;
+        }
+        /* assumes iov is uniformly on device or not */
+        cu_ret = cuPointerGetAttribute((void *) &local_ptr_ctx, attribute,
+                                       (CUdeviceptr) iov[0].buffer);
+        if (cu_ret != CUDA_SUCCESS) {
+            cuGetErrorString(cu_ret, &cu_err_str);
+            ucs_error("cuPointerGetAttribute failed ret:%s", cu_err_str);
+            return UCS_ERR_IO_ERROR;
+        }
+    }
+    if ((CUDA_SUCCESS == cu_err) && (key->dev_num == (int) cu_device) &&
```

Comment:
cu_err can be unintialized if cond at line 68 is false. I think instead of cu_err, have a variable  using_same_device_context=0/1

Reply:
Handled this.